docs: Debugging instructions using IDEs

bvrooman opened the pull request:
Related issues:
- Closes https://github.com/FuelLabs/fuel-core/issues/1236

This PR adds documentation on debugging.

Voxelot commented on docs/developers/debugging.md:
````diff
@@ -0,0 +1,131 @@
+cargo run --bin fuel-core-keygen new
+```
+
+The resulting key pair printed to the console contains an address and secret. The `address` can be provided to the `--keypair` argument when running the node.
````
```suggestion
The resulting key pair printed to the console contains an address and secret. The `secret` can be provided to the `--keypair` argument when running the node.
```

bvrooman replied:
Thanks for the clarification!

bvrooman commented on docs/developers/debugging.md:
```diff
@@ -0,0 +1,131 @@
+
+The resulting key pair printed to the console contains an address and secret. The `secret` can be provided to the `--keypair` argument when running the node.
+
+The `--network` argument identifies the name of the network to join. The network name is used during peer discovery. For example, users can specify `--network beta3-prod` to join the Beta 3 Production network. Similarly, setting the environment variable `NETWORK="beta3-prod"` will produce the same result.
```
Question just for my own understanding: If the names don't match, will the peers even discover each other? Is it not considered a separate network?

xgreenx replied:
Nope, they will not discover each other. The connection will be reset during authentication process before we started exchange data.

Voxelot commented:
minor conflict to resolve in the changelog, otherwise the docs look good

xgreenx reviewed (approved):
It would be nice to get some feedback from @luizstacio =)

xgreenx commented:
The issue also mentioned instructions on how to connect to the network. It's already done in https://github.com/FuelLabs/docs-hub/pull/36, so maybe you can add a link to our documentation instead.

bvrooman commented:
> It would be nice to get some feedback from @luizstacio =)

Certainly, it is always good to get more feedback. I will wait for any comments from Luiz

> The issue also mentioned instructions on how to connect to the network. It's already done in [FuelLabs/docs-hub#36](https://github.com/FuelLabs/docs-hub/pull/36), so maybe you can add a link to our documentation instead.

Thanks for pointing that out. I will have a look, and link the appropriate networking documentation.

luizstacio commented on docs/developers/debugging.md:
```diff
@@ -0,0 +1,133 @@
+
+1. In the top-right corner of the CLion window, click on the configurations dropdown menu and select "Edit Configurations..."
+2. In the "Run/Debug Configurations" window, click on the "+" button and select "Cargo"
+3. In the "Name" field, give your configuration a descriptive name, such as "Run Beta 3"
```
```suggestion
3. In the "Name" field, give your configuration a descriptive name, such as "Run Fuel Client"
```

Just to remove the need to update the docs. Maybe is better to set a generic name

bvrooman replied:
Yep, that makes sense 👍

bvrooman commented:
I updated the networking section with a link to the Fuel guide on Vercel. I removed some of the text explaining client networking, since the official documentation can do that better. I left the parts of the section that talk about configuration specifically, like setting the CLI argument or providing an environment variable.

luizstacio commented on docs/developers/debugging.md:
```diff
@@ -0,0 +1,133 @@
+
+### Example Configuration
+
+![clion_configuration.png](assets/clion_configuration.png)
```
Following the steps, I had to do;
5. Select the workspace folder
6. Remove on the before start section the build action

luizstacio replied:
Once I try to run it throws the error: 
```
ERROR fuel_core::cli: 57: Unable to load .env environment variables: path not found. Please check that you have created a .env file in your working directory.
```

bvrooman replied:
Thanks for going through the steps!

I think it would be a good idea for me to add a quick line adding the workspace folder. Usually, CLion should populate this line for you to your repository root (don't change it to what's exactly in the example!) But if that is missing for whatever reason, it would be helpful to know that you should set this to the root yourself. I see that I omitted that instruction.

The "Before start" settings are optional. CLion adds "Build" by default, but it's okay because `cargo run` does a build by default. I will add a note on these settings too.

bvrooman replied:
> Once I try to run it throws the error:
> 
> ```
> ERROR fuel_core::cli: 57: Unable to load .env environment variables: path not found. Please check that you have created a .env file in your working directory.
> ```

This is actually a new feature and it's deliberate. It simply means the client is expecting you to provide a `.env` file (probably because you compiled with `--all-features`), but it didn't find it. Simply add a `.env` file in the root of your repo with your specific configuration values, and you're good to go.

luizstacio replied:
Yeah, it worked even with the error. The real error was on the database folder: 

```
thread 'main' panicked at 'source slice length (64) does not match destination slice length (32)', <>/.cargo/registry/src/index.crates.io-6f17d22bba15001f/fuel-types-0.36.0/src/array_types.rs:389:16

luizstacio commented on docs/developers/debugging.md:
```diff
@@ -0,0 +1,133 @@
+2. VS Code will offer to create a `launch.json` file to configure how your client node is run. Clicking this will prompt you to select a debugger. Select the CodeLLDB debugger to proceed. VS Code will generate a `launch.json` file for you.
+3. In the "Run and Debug" menu, click the dropdown menu and select "Add Configuration..."
+4. In the `"type"` field, enter `"lldb"`. In the `"request"` field, enter `"launch"`.
+5. In the `"name"` field, give your configuration a descriptive name, such as `"Run Beta 3"`
```
```suggestion
5. In the `"name"` field, give your configuration a descriptive name, such as `"Run FuelCore"`
```

luizstacio commented on docs/developers/debugging.md:
```diff
@@ -0,0 +1,133 @@
+
+### Example Configuration
+
+![vs_code_configuration.png](assets/vs_code_configuration.png)
```
We could just past here the plain json, to make it easier to copy and paste

luizstacio replied:
I was not able to make the debug work, it looks like the command never finishes. It starts but not finishes.

My config;
```
[
    "run",
    "--all-features",
    "--bin",
    "fuel-core",
    "--",
    "run",
    "--ip", "0.0.0.0",
    "--port", "4004",
    "--db-type", "in-memory",
    "--utxo-validation",
    "--vm-backtrace",
    "--poa-interval-period",
    "30s"
]
```

bvrooman replied:
Luiz and I discussed this specific issue over a call: It appears that the client is starting correctly, and producing the expected output in the terminal, but the VS Code debugger is not correctly attaching to the running process. Luiz and I will spend some time independently looking into this.

bvrooman replied:
I updated the example with a more reliable config that users can copy + paste to get started.

luizstacio commented on docs/developers/debugging.md:
```diff
@@ -0,0 +1,133 @@
+1. In the top-right corner of the CLion window, click on the configurations dropdown menu and select "Edit Configurations..."
+2. In the "Run/Debug Configurations" window, click on the "+" button and select "Cargo"
+3. In the "Name" field, give your configuration a descriptive name, such as "Run Beta 3"
+4. In the "Command" field, enter `run --all-features --bin fuel-core -- run <ARGUMENTS>`, where `<ARGUMENTS>` represents the desired CLI arguments. Refer to the CLI documentation for guidance on which options and arguments to use.
```
I think we could add the example `--db-type in-memory` just to avoid errors.

bvrooman replied:
I added a Common Issues section to mention using `--db-type in-memory` in case of some errors.

xgreenx reviewed (approved):
Looks good to me, thank you!=)
